src/ripple/unl/tests/Sim3.h

```diff
@@ -0,0 +1,513 @@
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_UNL_SIM3_H_INCLUDED
+#define RIPPLE_UNL_SIM3_H_INCLUDED
+
+#include <ripple/unl/tests/BasicNetwork.h>
+#include <beast/container/aged_unordered_map.h>
+#include <boost/container/flat_set.hpp>
+#include <boost/optional.hpp>
+#include <algorithm>
+#include <memory>
+#include <sstream>
+#include <unordered_map>
+#include <vector>
+
+namespace ripple {
+namespace test {
+
+template <class Log>
+struct Sim3
+{
+    struct Config
+    {
+        int unl;
+        int peers = 100;
+        int trial = 100;
+    };
+
+    static int const nDegree = 10;      // outdegree
+    static int const nItem = 10;        // number of items
+    static int const nUpdateMS = 700;
+
+    using NodeKey = int;    // identifies a consensus participant
+    using ItemKey = int;    // identifies a ballot item
+    using ItemSet = boost::container::flat_set<ItemKey>;
+    using clock_type = std::chrono::system_clock;
+    using millis = std::chrono::milliseconds;
+
+    struct Network;
+
+    struct PosMsg
+    {
+        NodeKey id;
+        std::size_t seq;
+        ItemSet items;
+        bool last;
+    };
+
+    // A round of consensus.
+    // Each round consists of a series of votes,
+    // terminating when a supermajority is reached.
+    class Round
+    {
+    private:
+        int thresh_ = 50;
+
+    public:
+        struct Pos
+        {
+            ItemSet items;
+            bool last = false;
+            std::size_t seq = 0;
+        };
+
+        int id_;
+        bool failed_ = false;
+        bool consensus_ = false;
+        std::unordered_map<NodeKey, Pos> pos_;
+        std::size_t count_ = 0;
+        clock_type::time_point t0_;
+        Log& log_;
+
+    public:
+        // Create a new round with initial position
+        Round (NodeKey id, ItemSet&& pos,
+                clock_type::time_point now, Log& log)
+            : id_ (id)
+            , t0_ (now)
+            , log_ (log)
+        {
+            using namespace std;
+            pos_[id].items = std::move(pos);
+        }
+
+        std::shared_ptr<PosMsg const>
+        posMsg()
+        {
+            auto const iter = pos_.find(id_);
+            auto m = std::make_shared<PosMsg>();
```

Comment:
Looking at `Peer::broadcast` and `Peer::send`, I think we could get away with returning PosMsg by value and not allocating here at all (and rely on efficient moves to pass the parameter). The only sticking point is moving `m` into the lambda in `Peer::send`. We could write our own functor now, but I'd rather wait for our switch to C++-14 and generalized capture to do this. Just making a note.